velox/expression/VectorUdfTypeSystem.h

```diff
@@ -611,6 +616,88 @@ struct VectorWriter<ArrayProxyT<V>> {
   vector_size_t offset_ = 0;
 };
 
+// A new temporary vector writer, to be used when the user wants a writer
+// interface for the map output. Will eventually replace VectorWriter<Map>>.
+template <typename K, typename V>
+struct VectorWriter<MapWriterT<K, V>> {
+  using vector_t = typename TypeToFlatVector<Map<K, V>>::type;
+  using key_vector_t = typename TypeToFlatVector<K>::type;
+  using val_vector_t = typename TypeToFlatVector<V>::type;
+  using exec_out_t = MapWriter<K, V>;
+
+  void init(vector_t& vector) {
+    mapVector_ = &vector;
+    keyWriter_.init(static_cast<key_vector_t&>(*vector.mapKeys()));
+    valWriter_.init(static_cast<val_vector_t&>(*vector.mapValues()));
+    elementWriter_.initialize(this);
+  }
+
+  // This should be called once all rows are processed.
+  void finish() {
+    // Downsize to actual used size.
+    elementWriter_.keysVector_->resize(elementWriter_.innerOffset_);
+    elementWriter_.valuesVector_->resize(elementWriter_.innerOffset_);
+    mapVector_ = nullptr;
+  }
+
+  VectorWriter() = default;
+
+  exec_out_t& current() {
+    return elementWriter_;
+  }
+
+  vector_t& vector() {
+    return *mapVector_;
+  }
+
+  void ensureSize(size_t size) {
+    if (size > mapVector_->size()) {
+      mapVector_->resize(size);
+      init(vector());
+    }
+  }
+
+  // Commit a not null value.
+  void commit() {
+    mapVector_->setOffsetAndSize(
+        offset_, elementWriter_.innerOffset_, elementWriter_.length_);
+    mapVector_->setNull(offset_, false);
+    // Will reset length to 0 and prepare proxy_.valuesOffset_ for the next
+    // item.
+    elementWriter_.finalize();
+  }
+
+  // Commit a null value.
+  void commitNull() {
+    elementWriter_.finalizeNull();
```

Comment:
It'd be nice if there was a test that covered a case where some values were written, but ultimately null was committed

Reply:
test case added.